pkg/customresourcestate/registry_factory.go

```diff
@@ -0,0 +1,517 @@
+			result = append(result, *ev)
+		}
+	case []interface{}:
+		for i, it := range iter {
+			value, err := e.value(it)
+			if err != nil {
+				onError(fmt.Errorf("[%d]: %w", i, err))
+				continue
+			}
+			addPathLabels(it, e.LabelFromPath, value.Labels)
+			result = append(result, *value)
+		}
+	default:
+		value, err := e.value(v)
+		if err != nil {
+			onError(err)
+			break
+		}
+		addPathLabels(v, e.LabelFromPath, value.Labels)
+		result = append(result, *value)
+	}
+	return result, errors
+}
+
+func (e compiledEach) value(it interface{}) (*eachValue, error) {
+	labels := make(map[string]string)
+	value, err := getNum(e.ValueFrom.Get(it))
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", e.ValueFrom, err)
+	}
+	return &eachValue{
+		Labels: labels,
+		Value:  value,
+	}, nil
+}
+
+func (e eachValue) DefaultLabels(defaults map[string]string) {
+	for k, v := range defaults {
+		if _, ok := e.Labels[k]; !ok {
+			e.Labels[k] = v
+		}
+	}
+}
+func (e eachValue) ToMetric() *metric.Metric {
+	var keys, values []string
+	for k := range e.Labels {
+		keys = append(keys, k)
+	}
+	// make it deterministic
+	sort.Strings(keys)
+	for _, key := range keys {
+		values = append(values, e.Labels[key])
+	}
+	return &metric.Metric{
+		LabelKeys:   keys,
+		LabelValues: values,
+		Value:       e.Value,
+	}
+}
+
+type compiledFamily struct {
+	Name          string
+	Help          string
+	Each          compiledEach
+	Labels        map[string]string
+	LabelFromPath map[string]valuePath
+	ErrorLogV     klog.Level
+}
+
+func (f compiledFamily) BaseLabels(obj map[string]interface{}) map[string]string {
+	result := make(map[string]string)
+	for k, v := range f.Labels {
+		result[k] = v
+	}
+	addPathLabels(obj, f.LabelFromPath, result)
+	return result
+}
+
+func addPathLabels(obj interface{}, labels map[string]valuePath, result map[string]string) {
+	// *prefixed is a special case, it means copy an object
+	// always do that first so other labels can override
+	var stars []string
+	for k := range labels {
+		if strings.HasPrefix(k, "*") {
+			stars = append(stars, k)
+		}
+	}
+	sort.Strings(stars)
+	for _, k := range stars {
+		m := labels[k].Get(obj)
+		if kv, ok := m.(map[string]interface{}); ok {
+			for k, v := range kv {
+				result[k] = fmt.Sprintf("%v", v)
+			}
+		}
+	}
+	for k, v := range labels {
+		if strings.HasPrefix(k, "*") {
+			continue
+		}
+		result[k] = fmt.Sprintf("%v", v.Get(obj))
+	}
+}
+
+type pathOp struct {
+	part string
+	op   func(interface{}) interface{}
+}
+
+type valuePath []pathOp
+
+func (p valuePath) Get(obj interface{}) interface{} {
+	for _, op := range p {
+		if obj == nil {
+			return nil
+		}
+		obj = op.op(obj)
+	}
+	return obj
+}
+
+func (p valuePath) String() string {
+	var b strings.Builder
+	b.WriteRune('[')
+	for i, op := range p {
+		if i > 0 {
+			b.WriteRune(',')
+		}
+		b.WriteString(op.part)
+	}
+	b.WriteRune(']')
+	return b.String()
+}
+
+func compilePath(path []string) (out valuePath, _ error) {
```

Comment:
Can we use something like https://github.com/kubernetes/client-go/blob/master/util/jsonpath/jsonpath.go to reduce or replace this function?

Reply:
I attempted to drop it in, but there are a few problems:

* Bit of an impedance mismatch. jsonpath is expecting to parse a template string, and while I can get at the raw reflect values, a lot of "valid" expressions are going to result in output that doesn't make sense (multiple result values, etc.)
* Numeric handling is too strict. json.Unmarshal(..., map[string]any) parses all numerics as float64. So an expression like `$.Items[?(@.id==3)]` will never match anything because the types won't match. 

I don't love inventing a path expression syntax, but it's simple and flexible and gets the job done. JSONPath seems like overkill.